src/internal/_equals.js

```diff
@@ -1,114 +1,196 @@
+      return typeof a['fantasy-land/equals'] === 'function' && a['fantasy-land/equals'](b) &&
+        typeof b['fantasy-land/equals'] === 'function' && b['fantasy-land/equals'](a);
+    }
+
+    if (typeof a.equals === 'function' || typeof b.equals === 'function') {
+      return typeof a.equals === 'function' && a.equals(b) &&
+        typeof b.equals === 'function' && b.equals(a);
+    }
+
+    switch (type(a)) {
+      case 'Arguments':
+      case 'Array':
+      case 'Object':
+        if (typeof a.constructor === 'function' &&
           _functionName(a.constructor) === 'Promise') {
-        return a === b;
-      }
-      break;
-    case 'Boolean':
-    case 'Number':
-    case 'String':
-      if (!(typeof a === typeof b && identical(a.valueOf(), b.valueOf()))) {
-        return false;
-      }
-      break;
-    case 'Date':
-      if (!identical(a.valueOf(), b.valueOf())) {
-        return false;
-      }
-      break;
-    case 'Error':
-      return a.name === b.name && a.message === b.message;
-    case 'RegExp':
-      if (!(a.source === b.source &&
+          return a === b;
+        }
+        break;
+      case 'Boolean':
+      case 'Number':
+      case 'String':
+        if (!(typeof a === typeof b && identical(a.valueOf(), b.valueOf()))) {
+          return false;
+        }
+        break;
+      case 'Date':
+        if (!identical(a.valueOf(), b.valueOf())) {
+          return false;
+        }
+        break;
+      case 'Error':
+        return a.name === b.name && a.message === b.message;
+      case 'RegExp':
+        if (!(a.source === b.source &&
             a.global === b.global &&
             a.ignoreCase === b.ignoreCase &&
             a.multiline === b.multiline &&
             a.sticky === b.sticky &&
             a.unicode === b.unicode)) {
-        return false;
-      }
-      break;
-    case 'Map':
-    case 'Set':
-      if (!_equals(_arrayFromIterator(a.entries()), _arrayFromIterator(b.entries()), stackA, stackB)) {
-        return false;
-      }
-      break;
-    case 'Int8Array':
-    case 'Uint8Array':
-    case 'Uint8ClampedArray':
-    case 'Int16Array':
-    case 'Uint16Array':
-    case 'Int32Array':
-    case 'Uint32Array':
-    case 'Float32Array':
-    case 'Float64Array':
-      break;
-    case 'ArrayBuffer':
-      break;
-    default:
-      // Values of other types are only equal if identical.
-      return false;
-  }
+          return false;
+        }
+        break;
+      case 'Map':
+        if (a.size !== b.size) {
+          return false;
+        }
 
-  var keysA = keys(a);
-  if (keysA.length !== keys(b).length) {
-    return false;
-  }
+        var stackEquals = _stackEquals(a, b, stackA, stackB);
+        if (stackEquals !== undefined) {
+          return stackEquals;
+        }
+
+        stackA.push(a);
+        stackB.push(b);
+
+        // this optimization can be not needed, depends on use case.
```

Comment:
I'm not sure. It is performance-related question.
This is the thing to discuss: on one hand, in case we're checking keys first, we're spending **n** time on equality check first, and then, if all keys are same, we're spending around **2n** time to validate KVs. So, if keys are same, we're spending **n** time on it, and **3n** otherwise. On other hand, if we're simply checking the KVs, we're spending around **2n** time anyway.

On other hand - if we're talking about the small Maps, we can consider that wrapping code working time is nearly same as inside code, and for big maps - the bigger the maps are, the less is the chance we're talking about 2 maps of same length and different keys (general collision case is "let's have 2 different maps for same values to store different whatever, cache, for example")